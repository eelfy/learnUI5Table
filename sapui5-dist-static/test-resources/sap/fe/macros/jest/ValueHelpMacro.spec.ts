import ValueHelpMetadata from "sap/fe/macros/ValueHelp.metadata";
import { compileCDS, getTemplatingResult, registerMacro, serializeXML, getControlAttribute } from "sap/fe/test/JestTemplatingHelper";
import path from "path";

describe("Value Help templating", () => {
	beforeEach(() => {
		registerMacro(ValueHelpMetadata);
	});

	const sMetadataUrl = compileCDS(path.join(__dirname, "./data/ValueHelpMacro.cds"));
	const mBindingContexts = {
		"entitySet": "/Items"
	};

	describe("- check control templating - ", () => {
		it("Properties without value help and outside the filter bar (no condition panel) ", async () => {
			let xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									property="ID"
									/>`;

			let domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(domResult).toNotHaveControl("/mdcField:FieldValueHelp");

			xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									property="stringValue"
									/>`;

			domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(domResult).toNotHaveControl("/mdcField:FieldValueHelp");

			xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									property="decimalValue"
									/>`;

			domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(domResult).toNotHaveControl("/mdcField:FieldValueHelp");
		});

		it("Properties of type date will have a value help ", async () => {
			const xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									property="decimalValue"
									/>`;

			const domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(domResult).toNotHaveControl("/mdcField:FieldValueHelp");
			expect(serializeXML(domResult)).toMatchSnapshot();
		});

		it("Properties with unit will have a value help on that unit ", async () => {
			const xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									property="quantity"
									/>`;

			const domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(domResult).toHaveControl("/mdcField:FieldValueHelp");
			expect(serializeXML(domResult)).toMatchSnapshot();
		});

		it("Properties with unit will have not have value help on that unit in the filter field ", async () => {
			const xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									property="quantity"
									filterFieldValueHelp="true"
									forceValueHelp="true"
									/>`;

			const domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(domResult).toHaveControl("/mdcField:FieldValueHelp");
			expect(getControlAttribute("/mdcField:FieldValueHelp", "id", domResult)).toEqual("ValueHelp::quantity");
			expect(getControlAttribute("/mdcField:FieldValueHelp", "showConditionPanel", domResult)).toEqual("true");
			expect(serializeXML(domResult)).toMatchSnapshot();
		});

		it("Value with value help", async () => {
			const xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									idPrefix="VH"
									property="vhReference"
									conditionModel="$conditions"
									filterFieldValueHelp="false"
									requestGroupId="Yolo"
									/>`;

			const domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(serializeXML(domResult)).toMatchSnapshot();

			const xmlFieldInFilterField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									idPrefix="VH"
									property="vhReference"
									conditionModel="$conditions"
									filterFieldValueHelp="false"
									requestGroupId="Yolo"
									/>`;

			const domResultInFilterField = await getTemplatingResult(xmlFieldInFilterField, sMetadataUrl, mBindingContexts, {});
			expect(serializeXML(domResultInFilterField)).toMatchSnapshot();
			expect(serializeXML(domResultInFilterField)).toEqual(serializeXML(domResult));
		});

		it("Value with value help with fixed values", async () => {
			const xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									idPrefix="VH"
									property="vhReferenceWithFixedValues"
									conditionModel="$conditions"
									filterFieldValueHelp="false"
									requestGroupId="Yolo"
									/>`;

			const domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(serializeXML(domResult)).toMatchSnapshot();
		});

		it("Value with value help following navproperty", async () => {
			const xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									idPrefix="VH"
									property="toSubItems/subvhReference"
									conditionModel="$conditions"
									filterFieldValueHelp="false"
									requestGroupId="Yolo"
									/>`;

			const domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(serializeXML(domResult)).toMatchSnapshot();
			expect(getControlAttribute("/mdcField:FieldValueHelp", "id", domResult)).toEqual("VH::toSubItems::subvhReference");
		});

		it("Value with value help following navproperty", async () => {
			const xmlField = `<macro:ValueHelp xmlns:macro="sap.fe.macros"
									_flexId="MyFlexID"
									property="toSubItems/subvhReference"
									conditionModel="$conditions"
									filterFieldValueHelp="false"
									requestGroupId="Yolo"
									/>`;

			const domResult = await getTemplatingResult(xmlField, sMetadataUrl, mBindingContexts, {});
			expect(serializeXML(domResult)).toMatchSnapshot();
			expect(getControlAttribute("/mdcField:FieldValueHelp", "id", domResult)).toEqual("MyFlexID");
		});
	});
});
