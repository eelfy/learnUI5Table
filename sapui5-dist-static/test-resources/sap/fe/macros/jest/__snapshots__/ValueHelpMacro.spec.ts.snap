// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value Help templating - check control templating -  Properties of type date will have a value help  1`] = `
"<root> </root>
"
`;

exports[`Value Help templating - check control templating -  Properties with unit will have a value help on that unit  1`] = `
"<root>
  <mdcField:FieldValueHelp
    xmlns:mdcField=\\"sap.ui.mdc.field\\"
    xmlns:ns1=\\"sap.ui.core\\"
    ns1:require=\\"{ValueListHelper: 'sap/fe/macros/internal/valuehelp/ValueListHelper'}\\"
    xmlns:ns2=\\"http://schemas.sap.com/sapui5/preprocessorextension/sap.fe.unittesting/1\\"
    ns2:id=\\"ValueHelpMacroFragment\\"
    delegate=\\"{name: 'sap/fe/macros/FieldValueHelpDelegate', payload: {propertyPath: '/Items/quantityUnit', conditionModel: ''}}\\"
    id=\\"ValueHelp::quantity::quantityUnit\\"
    showConditionPanel=\\"false\\"
    validateInput=\\"false\\"
    noDialog=\\"false\\"
    xmlns:ns3=\\"http://schemas.sap.com/sapui5/extension/sap.ui.core.CustomData/1\\"
    ns3:requestGroupId=\\"\\"
    ns3:sourcePath=\\"/Items/quantityUnit\\"
  >
    <mdcField:FieldValueHelpMTableWrapper />
  </mdcField:FieldValueHelp>
</root>
"
`;

exports[`Value Help templating - check control templating -  Properties with unit will have not have value help on that unit in the filter field  1`] = `
"<root>
  <mdcField:FieldValueHelp
    xmlns:mdcField=\\"sap.ui.mdc.field\\"
    xmlns:ns1=\\"http://schemas.sap.com/sapui5/preprocessorextension/sap.fe.unittesting/1\\"
    ns1:id=\\"ValueHelpMacroFragmentWithDateTimeFilters\\"
    id=\\"ValueHelp::quantity\\"
    showConditionPanel=\\"true\\"
    title=\\"\\"
    xmlns:ns2=\\"http://schemas.sap.com/sapui5/extension/sap.ui.core.CustomData/1\\"
    ns2:requestGroupId=\\"\\"
  >
  </mdcField:FieldValueHelp>
</root>
"
`;

exports[`Value Help templating - check control templating -  Value with value help 1`] = `
"<root>
  <mdcField:FieldValueHelp
    xmlns:mdcField=\\"sap.ui.mdc.field\\"
    xmlns:ns1=\\"sap.ui.core\\"
    ns1:require=\\"{ValueListHelper: 'sap/fe/macros/internal/valuehelp/ValueListHelper'}\\"
    xmlns:ns2=\\"http://schemas.sap.com/sapui5/preprocessorextension/sap.fe.unittesting/1\\"
    ns2:id=\\"ValueHelpMacroFragment\\"
    delegate=\\"{name: 'sap/fe/macros/FieldValueHelpDelegate', payload: {propertyPath: '/Items/vhReference', conditionModel: '$conditions'}}\\"
    id=\\"VH::vhReference\\"
    showConditionPanel=\\"false\\"
    validateInput=\\"false\\"
    noDialog=\\"false\\"
    xmlns:ns3=\\"http://schemas.sap.com/sapui5/extension/sap.ui.core.CustomData/1\\"
    ns3:requestGroupId=\\"Yolo\\"
    ns3:sourcePath=\\"/Items/vhReference\\"
  >
    <mdcField:FieldValueHelpMTableWrapper />
  </mdcField:FieldValueHelp>
</root>
"
`;

exports[`Value Help templating - check control templating -  Value with value help 2`] = `
"<root>
  <mdcField:FieldValueHelp
    xmlns:mdcField=\\"sap.ui.mdc.field\\"
    xmlns:ns1=\\"sap.ui.core\\"
    ns1:require=\\"{ValueListHelper: 'sap/fe/macros/internal/valuehelp/ValueListHelper'}\\"
    xmlns:ns2=\\"http://schemas.sap.com/sapui5/preprocessorextension/sap.fe.unittesting/1\\"
    ns2:id=\\"ValueHelpMacroFragment\\"
    delegate=\\"{name: 'sap/fe/macros/FieldValueHelpDelegate', payload: {propertyPath: '/Items/vhReference', conditionModel: '$conditions'}}\\"
    id=\\"VH::vhReference\\"
    showConditionPanel=\\"false\\"
    validateInput=\\"false\\"
    noDialog=\\"false\\"
    xmlns:ns3=\\"http://schemas.sap.com/sapui5/extension/sap.ui.core.CustomData/1\\"
    ns3:requestGroupId=\\"Yolo\\"
    ns3:sourcePath=\\"/Items/vhReference\\"
  >
    <mdcField:FieldValueHelpMTableWrapper />
  </mdcField:FieldValueHelp>
</root>
"
`;

exports[`Value Help templating - check control templating -  Value with value help following navproperty 1`] = `
"<root>
  <mdcField:FieldValueHelp
    xmlns:mdcField=\\"sap.ui.mdc.field\\"
    xmlns:ns1=\\"sap.ui.core\\"
    ns1:require=\\"{ValueListHelper: 'sap/fe/macros/internal/valuehelp/ValueListHelper'}\\"
    xmlns:ns2=\\"http://schemas.sap.com/sapui5/preprocessorextension/sap.fe.unittesting/1\\"
    ns2:id=\\"ValueHelpMacroFragment\\"
    delegate=\\"{name: 'sap/fe/macros/FieldValueHelpDelegate', payload: {propertyPath: '/Items/toSubItems/subvhReference', conditionModel: '$conditions'}}\\"
    id=\\"VH::toSubItems::subvhReference\\"
    showConditionPanel=\\"false\\"
    validateInput=\\"false\\"
    noDialog=\\"false\\"
    xmlns:ns3=\\"http://schemas.sap.com/sapui5/extension/sap.ui.core.CustomData/1\\"
    ns3:requestGroupId=\\"Yolo\\"
    ns3:sourcePath=\\"/Items/toSubItems/subvhReference\\"
  >
    <mdcField:FieldValueHelpMTableWrapper />
  </mdcField:FieldValueHelp>
</root>
"
`;

exports[`Value Help templating - check control templating -  Value with value help following navproperty 2`] = `
"<root>
  <mdcField:FieldValueHelp
    xmlns:mdcField=\\"sap.ui.mdc.field\\"
    xmlns:ns1=\\"sap.ui.core\\"
    ns1:require=\\"{ValueListHelper: 'sap/fe/macros/internal/valuehelp/ValueListHelper'}\\"
    xmlns:ns2=\\"http://schemas.sap.com/sapui5/preprocessorextension/sap.fe.unittesting/1\\"
    ns2:id=\\"ValueHelpMacroFragment\\"
    delegate=\\"{name: 'sap/fe/macros/FieldValueHelpDelegate', payload: {propertyPath: '/Items/toSubItems/subvhReference', conditionModel: '$conditions'}}\\"
    id=\\"MyFlexID\\"
    showConditionPanel=\\"false\\"
    validateInput=\\"false\\"
    noDialog=\\"false\\"
    xmlns:ns3=\\"http://schemas.sap.com/sapui5/extension/sap.ui.core.CustomData/1\\"
    ns3:requestGroupId=\\"Yolo\\"
    ns3:sourcePath=\\"/Items/toSubItems/subvhReference\\"
  >
    <mdcField:FieldValueHelpMTableWrapper />
  </mdcField:FieldValueHelp>
</root>
"
`;

exports[`Value Help templating - check control templating -  Value with value help with fixed values 1`] = `
"<root>
  <mdcField:FieldValueHelp
    xmlns:mdcField=\\"sap.ui.mdc.field\\"
    xmlns:ns1=\\"sap.ui.core\\"
    ns1:require=\\"{ValueListHelper: 'sap/fe/macros/internal/valuehelp/ValueListHelper'}\\"
    xmlns:ns2=\\"http://schemas.sap.com/sapui5/preprocessorextension/sap.fe.unittesting/1\\"
    ns2:id=\\"ValueHelpMacroFragment\\"
    delegate=\\"{name: 'sap/fe/macros/FieldValueHelpDelegate', payload: {propertyPath: '/Items/vhReferenceWithFixedValues', conditionModel: '$conditions'}}\\"
    id=\\"VH::vhReferenceWithFixedValues\\"
    showConditionPanel=\\"false\\"
    validateInput=\\"true\\"
    noDialog=\\"true\\"
    xmlns:ns3=\\"http://schemas.sap.com/sapui5/extension/sap.ui.core.CustomData/1\\"
    ns3:requestGroupId=\\"Yolo\\"
    ns3:sourcePath=\\"/Items/vhReferenceWithFixedValues\\"
  >
    <mdcField:FieldValueHelpMTableWrapper />
  </mdcField:FieldValueHelp>
</root>
"
`;
